sql/pipelines/src/main/scala/org/apache/spark/sql/pipelines/graph/FlowExecution.scala

```diff
@@ -0,0 +1,256 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.pipelines.graph
+
+import java.util.concurrent.ThreadPoolExecutor
+import java.util.concurrent.atomic.AtomicBoolean
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+import scala.util.control.NonFatal
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.classic.SparkSession
+import org.apache.spark.sql.pipelines.graph.QueryOrigin.ExceptionHelpers
+import org.apache.spark.sql.pipelines.util.SparkSessionUtils
+import org.apache.spark.sql.streaming.{OutputMode, StreamingQuery, Trigger}
+import org.apache.spark.util.ThreadUtils
+
+/**
+ * A flow's execution may complete for two reasons:
+ * 1. it may finish performing all of its necessary work, or
+ * 2. it may be interrupted by a request from a user to stop it.
+ *
+ * We use this result to disambiguate these two cases, using [[ExecutionResult.FINISHED]]
+ * for the former and [[ExecutionResult.STOPPED]] for the latter.
+ */
+sealed trait ExecutionResult
+object ExecutionResult {
+  case object FINISHED extends ExecutionResult
+  case object STOPPED extends ExecutionResult
+}
+
+/** A [[FlowExecution]] specifies how to execute a flow and manages its execution. */
+trait FlowExecution {
```

Comment:
each logical dataflow (DataflowGraph::Flow) would be translated into a concrete flow execution plan (FlowExecution) when we start the pipeline execution (in FlowPlanner.scala).

Here we define the interface for FlowExecution